code/workspaces/infrastructure-api/src/controllers/projectsController.js

```diff
@@ -0,0 +1,108 @@
+import { check, matchedData } from 'express-validator';
+import { service } from 'common';
+import projectsRepository from '../dataaccess/projectsRepository';
+
+async function listProjects(request, response, next) {
+  try {
+    const projects = await projectsRepository.getAll();
+    response.send(projects);
+  } catch (error) {
+    next(new Error(`Error listing projects: ${error.message}`));
+  }
+}
+
+async function getProjectByKey(request, response, next) {
+  const { projectKey } = matchedData(request);
+
+  try {
+    const project = await projectsRepository.getByKey(projectKey);
+    if (project) {
+      response.send(project);
+    } else {
+      response.status(404).send();
+    }
+  } catch (error) {
+    next(new Error(`Error getting project by key: ${error.message}`));
+  }
+}
+
+async function createProject(request, response, next) {
+  const project = matchedData(request, { locations: ['body'] });
+
+  if (await projectsRepository.exists(project.projectKey)) {
+    response.status(400).send({
+      errors: [{ msg: `Entry with projectKey '${project.projectKey}' already exists.` }],
+    });
+  } else {
+    try {
+      const createdProject = await projectsRepository.create(project);
+      response.status(201).send(createdProject);
+    } catch (error) {
+      next(new Error(`Error creating project: ${error.message}`));
+    }
+  }
+}
+
+async function createOrUpdateProject(request, response, next) {
+  const { projectKey } = matchedData(request, { locations: ['params'] });
+  const project = matchedData(request, { locations: ['body'] });
+
+  if (projectKey !== project.projectKey) {
+    const msg = "'projectKey' parameter in URL and 'projectKey' field in request body must "
+      + `match. Got '${projectKey}' in URL and '${project.projectKey}' in body.`;
+    response.status(400).send({ errors: [{ msg }] });
+  } else {
+    try {
+      const exists = await projectsRepository.exists(projectKey);
```

Comment:
Not sure it matters, but could we find whether we created or inserted from the upsert response? I feel like that is what the infra watcher service does.

Reply:
When I was making the update I couldn't see anything in the Mongoose documentation. In the infra watcher the function used is `updateOne` rather than `findOneAndUpdate`. The former returns information about how many have been found and updated and the second returns the updated document so unable to workout whether it has been created or updated. Therefore, I will leave this code as it is.